client/web/web.go

```diff
@@ -371,6 +434,88 @@ func (s *Server) serveTailscaleAuth(w http.ResponseWriter, r *http.Request) {
 	w.Header().Set("Content-Type", "application/json")
 }
 
+func (s *Server) newSessionCookie() (string, error) {
+	raw := make([]byte, 16)
+	for i := 0; i < 5; i++ {
+		if _, err := rand.Read(raw); err != nil {
+			return "", err
+		}
+		cookie := "ts-web-" + base64.RawURLEncoding.EncodeToString(raw)
+		if _, ok := s.browserSessions.Load(cookie); !ok {
+			return cookie, nil
+		}
+	}
+	return "", errors.New("too many collisions generating new session; please refresh page")
+}
+
+func (s *Server) getControlServerURL(ctx context.Context) (string, error) {
+	if v := s.controlServerURL.Load(); v != nil {
+		v, _ := v.(string)
+		return v, nil
+	}
+	prefs, err := s.lc.GetPrefs(ctx)
+	if err != nil {
+		return "", err
+	}
+	url := prefs.ControlURLOrDefault()
+	s.controlServerURL.Store(url)
+	return url, nil
+}
+
+// getOrAwaitAuthURL connects to the control server for user auth,
+// with the following behavior:
+//
+//  1. If authURL is provided empty, a new auth URL is created on the
+//     control server and reported back here, which can then be used
+//     to redirect the user on the frontend.
+//  2. If authURL is provided non-empty, the connection to control
+//     blocks until the user has completed the URL. getOrAwaitAuthURL
+//     terminates when either the URL is completed, or ctx is canceled.
+func (s *Server) getOrAwaitAuthURL(ctx context.Context, authURL string, src tailcfg.NodeID) (*tailcfg.WebClientAuthResponse, error) {
+	serverURL, err := s.getControlServerURL(ctx)
+	if err != nil {
+		return nil, err
+	}
+	type data struct {
+		ID  string
+		Src tailcfg.NodeID
+	}
+	var b bytes.Buffer
+	if err := json.NewEncoder(&b).Encode(data{
+		ID:  strings.TrimPrefix(authURL, serverURL),
+		Src: src,
+	}); err != nil {
+		return nil, err
+	}
+	url := "http://" + apitype.LocalAPIHost + "/localapi/v0/debug-web-client"
+	req, err := http.NewRequestWithContext(ctx, "POST", url, &b)
+	if err != nil {
+		return nil, err
+	}
+	resp, err := s.lc.DoLocalRequest(req)
+	if err != nil {
+		return nil, err
+	}
+	body, _ := io.ReadAll(resp.Body)
+	resp.Body.Close()
+	if resp.StatusCode == http.StatusUnauthorized {
+		// User completed auth, but control server deemed
+		// them unauthorized to manage this node.
+		//
+		// TODO(sonia,will): not doing anything with the
```

Comment:
The only time control returns unauthorized from the webclient auth endpoint is if we weren't able to authenticate the noise request... so they machine key doesn't own the dst node key, the source node key has expired (which *should* prevent the connect to dst in the first place), or the source node key was revoked (again, that should prevent the initial connection).  Control doesn't make any determination about whether the authenticated user should be able to manage the dst node, it just authenticates them.

so the code is fine, I just don't think we'll ever want to surface these error messages to the frontend... they should be extraordinarily rare, and there's nothing they can do but retry.

Reply:
Also throws this error if they accidentally log in as the wrong user on control? But yeah agreed - I'll remove this TODO.

Reply:
uh.... does it?  Looking at the code, it appears that it would return a 200 with a AuthResponse error of "Auth Failed"?  That's probably not what we want if we're actually using proper status codes everywhere else.

Reply:
Oh yeah you're right. Hmm yeah it's weird because the auth path was complete, but auth failed. I do think maybe we should return error instead?